File: dispatch/java/adamma/c4dhi/claid/Module/Scheduling/RunnableDispatcher.java
package adamma.c4dhi.claid.Module.Scheduling;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import adamma.c4dhi.claid.Logger.Logger;

import java.util.concurrent.TimeUnit;
import java.util.Map;
import java.util.TreeMap;
import java.util.List;
import java.sql.Time;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;

public class RunnableDispatcher 
{
    private Map<LocalDateTime, ScheduledRunnable> scheduledRunnables = new TreeMap<>();
    private final Lock mutex = new ReentrantLock();
    private final Condition conditionVariable = mutex.newCondition();
    private boolean rescheduleRequired = false;
    private boolean running = false;

    private Thread scheduleThread;

    private long localDateTimeToMilliseconds(LocalDateTime localDateTime)
    {
        long milliseconds = localDateTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
        return milliseconds;
    }

    private long getWaitDurationUntilNextRunnableIsDue() 
    {
        mutex.lock();
        try 
        {
            if (scheduledRunnables.isEmpty()) 
            {
                // Wait forever.
                return Long.MAX_VALUE;
            }

            long now = System.currentTimeMillis();
            LocalDateTime nextExecutionTime = scheduledRunnables.keySet().iterator().next();
            long timeUntilNextRunnableIsDue = localDateTimeToMilliseconds(nextExecutionTime) - now;

            // Make sure we do not return a negative duration.
            return timeUntilNextRunnableIsDue < 0 ? 0 : timeUntilNextRunnableIsDue;
        } 
        finally 
        {
            mutex.unlock();
        }
    }

    private void waitUntilRunnableIsDueOrRescheduleIsRequired() 
    {

        long waitTime = getWaitDurationUntilNextRunnableIsDue();



        mutex.lock();
        try {
            conditionVariable.await(waitTime, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } finally {
            mutex.unlock();
        }
    }

    private void processRunnable(ScheduledRunnable scheduledRunnable) {
        if (!running) {
            return;
        }

        if (scheduledRunnable.isValid()) 
        {
            Logger.logInfo("Running runnable");

            scheduledRunnable.run();
            
            Logger.logInfo("Done running runnalbe");


            /*if (scheduledRunnable.runnable.stopDispatcherAfterThisRunnable) {
                // No more runnables will be executed after this one!
                stop();
                System.out.println("STOPPED DISPATCHER!");
                return;
            }*/

            if (scheduledRunnable.schedule.doesRunnableHaveToBeRepeated())
             {
                mutex.lock();
                try {

                    scheduledRunnable.schedule.updateExecutionTime();

                    LocalDateTime newExecutionTime = scheduledRunnable.schedule.getExecutionTime();

                    // Reinsert the runnable with the new scheduled execution time.
                    // Note that the runnable was removed from scheduledRunnables in the getAndRemoveDueRunnables() function.
                    scheduledRunnables.put(newExecutionTime, scheduledRunnable);
                } finally {
                    mutex.unlock();
                }
            }
        }
        else
        {
            Logger.logInfo("Runnable not valid.");
        }
    }

    private void processRunnables(List<ScheduledRunnable> runnables) {

        for (ScheduledRunnable runnable : runnables) {
            processRunnable(runnable);
        }
    }

    private List<ScheduledRunnable> getAndRemoveDueRunnables() {
        List<ScheduledRunnable> runnables = new ArrayList<>();
        long now = System.currentTimeMillis();

        mutex.lock();
        try {
            List<LocalDateTime> keysToRemove = new ArrayList<>();

            for (Map.Entry<LocalDateTime, ScheduledRunnable> entry : scheduledRunnables.entrySet()) {
                LocalDateTime dueTime = entry.getKey();

                long dueTimeMillis = localDateTimeToMilliseconds(dueTime);
                if (now >= dueTimeMillis) {
                    runnables.add(entry.getValue());
                    keysToRemove.add(dueTime);
                } else {
                    // Since the map is ordered in ascending order,
                    // all subsequent entries have a time
                    // which is greater than the time of the current entry.
                    // Hence, no need to keep iterating. We found the latest-most runnable
                    // which is due.
                    break;
                }
            }

            // If at least one runnable was added to runnables, that means we have to remove them from scheduledRunnables.
            for (LocalDateTime key : keysToRemove) {
                scheduledRunnables.remove(key);
            }
        } finally {
            mutex.unlock();
        }

        return runnables;
    }

    public boolean start() 
    {
        if(running)
        {
            return false;
        }
        running = true;
        scheduleThread = new Thread()
        {
            public void run()
            {
                runScheduling();
            }
        };
        scheduleThread.start();
        return true;
    }

    public boolean stop() {
        if(!running)
        {
            return false;
        }
        running = false;

        return true;
    }

    public void addRunnable(ScheduledRunnable runnable) {
        mutex.lock();
        try {
            LocalDateTime executionTime = runnable.schedule.getExecutionTime();

            while(scheduledRunnables.containsKey(executionTime))
            {
                    Logger.logInfo("Runnables contains key");

                // Reschedule the runnable a bit later.
                // Plus 1000 ns = plus 1 ms
                executionTime = executionTime.plusNanos(1000);
            }

            scheduledRunnables.put(executionTime, runnable);
                Logger.logInfo("Runnables " + scheduledRunnables.size());
            // This will lead to a wake-up, so we can reschedule.
            rescheduleRequired = true;
            conditionVariable.signalAll();
        } finally {
            mutex.unlock();
        }
    }

    public void runScheduling() {
        List<ScheduledRunnable> dueRunnables = new ArrayList<>();
        while (running) {
            do {
                Logger.logInfo("Scheduling " );
                // While we process runnables, it is possible
                // that another runnable becomes due in the meantime.
                // Hence, we repeat this loop until there are no more
                // runnables that are due.
                dueRunnables = getAndRemoveDueRunnables();
                processRunnables(dueRunnables);
            } while (!dueRunnables.isEmpty());
            rescheduleRequired = false;
            waitUntilRunnableIsDueOrRescheduleIsRequired();
        }
    }
}

File: dispatch/python/claid/module/scheduling/runnable_dispatcher.py
###########################################################################
# Copyright (C) 2023 ETH Zurich
# CLAID: Closing the Loop on AI & Data Collection (https://claid.ethz.ch)
# Core AI & Digital Biomarker, Acoustic and Inflammatory Biomarkers (ADAMMA)
# Centre for Digital Health Interventions (c4dhi.org)
# 
# Authors: Patrick Langer
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#         http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

import threading
import time
from typing import List
from logger.logger import Logger
from datetime import datetime
import asyncio

class RunnableDispatcher:
    def __init__(self):
        # Data structures
        self.scheduled_runnables = []

        # Lock & condition for asyncio
        self.lock = asyncio.Lock()
        # Workaround for issue #358: https://github.com/aio-libs/janus/issues/358
        getattr(self.lock, '_get_loop', lambda: None)()
        self.condition = asyncio.Condition(self.lock)

        self.reschedule_required = False
        self.stopped = True

        # Instead of a Thread, we'll store an asyncio Task
        self._scheduler_task = None


    async def start(self):
        """Start the dispatcher as an async Task."""
        async with self.lock:
            if not self.stopped:
                return False
            self.stopped = False

        # Create and schedule the run_scheduling() task
        self._scheduler_task = asyncio.create_task(self.run_scheduling())
        return True

    async def stop(self):
        """Stop the dispatcher and wait for the scheduling task to finish."""
        async with self.lock:
            if self.stopped:
                return False
            self.stopped = True
            # Wake up the run_scheduling() loop
            self.condition.notify_all()

        # Wait for the scheduling task to exit
        if self._scheduler_task:
            await self._scheduler_task

        return True

    async def add_runnable(self, scheduled_runnable):
        """Add a runnable and notify the dispatcher to possibly reschedule."""
        Logger.log_info("add runnable 1")
        async with self.lock:
            Logger.log_info("add runnable 2")
            self.scheduled_runnables.append(scheduled_runnable)
            self.scheduled_runnables.sort(key=lambda x: x.schedule.get_execution_time())
            Logger.log_info(
                f"Added runnable, total runnables now: {len(self.scheduled_runnables)}"
            )
            Logger.log_info("add runnable 3")

            self.reschedule_required = True
            Logger.log_info(f"{self} Notified")
            self.condition.notify_all()  # Wake the loop so it can reschedule

    def get_wait_duration_until_next_runnable_is_due(self):
        """
        Computes how long (in milliseconds) until the next runnable is due.
        This is a synchronous check of self.scheduled_runnables,
        but it must be called while the lock is held.
        """
        if not self.scheduled_runnables:
            # Nothing is scheduled, so wait a very long time.
            return 1_000_000.0  # effectively "forever"

        next_due_time = self.scheduled_runnables[0].schedule.get_execution_time()
        diff_seconds = (next_due_time - datetime.now()).total_seconds()
        milliseconds_until_due = diff_seconds * 1000.0
        return max(0.0, milliseconds_until_due)  # No negative durations

    async def _wait_for_reschedule(self):
        """Wait until `self.reschedule_required` becomes True."""
        async with self.condition:
            while not self.reschedule_required and not self.stopped:
                Logger.log_info("Waiting")
                await self.condition.wait()  # Wait for notification
                Logger.log_info(f"Waiting done {self.reschedule_required} {self.stopped}")

    async def wait_until_runnable_is_due_or_reschedule_is_required(self):
        """
        Wait until either:
          - The next runnable is due (timeout expires), or
          - self.reschedule_required or self.stopped is set to True.
        """
        Logger.log_info("Getting wait time")
        async with self.lock:
            # Compute how long to wait (in seconds for asyncio)
            wait_time_in_seconds = self.get_wait_duration_until_next_runnable_is_due() / 1000.0
        Logger.log_info("Getting wait time2")

        # Wait for either the predicate or a timeout.
        # Note: async with self.condition *again*, because wait_for needs the condition locked.
        try:
            Logger.log_info("Getting wait time")

          
            Logger.log_info("Getting wait time 2")
            await asyncio.wait_for(
                self._wait_for_reschedule(),
                timeout=wait_time_in_seconds
            )
            Logger.log_info("Condition woke up")
        except asyncio.TimeoutError:
            # A TimeoutError just means a new runnable was added
            # or we waited for timeout. We can safely ignore the "error".
            Logger.log_info("Condition woke up")
            pass

    async def run_scheduling(self):
        """The main scheduling loop that runs until stopped."""
        Logger.log_info("Running scheduling")
        while True:
            async with self.lock:
                if self.stopped:
                    break

            # Gather runnables that are due NOW
            due_runnables = await self.get_and_remove_due_runnables()
            while due_runnables:
                # Process them
                await self.process_runnables(due_runnables)
                # Check if more became due while we were processing
                due_runnables = await self.get_and_remove_due_runnables()
            Logger.log_info("done processing")
            # Reset reschedule_required before we wait again
            async with self.lock:
                self.reschedule_required = False

            # Now wait for the next due runnable or a reschedule
            await self.wait_until_runnable_is_due_or_reschedule_is_required()

        Logger.log_info("RunnableDispatcher shutdown.")

    async def get_and_remove_due_runnables(self):
        """Return and remove all runnables that are currently due."""
        async with self.lock:
            runnables = []
            now = datetime.now()
            while self.scheduled_runnables:
                due_runnable = self.scheduled_runnables[0]
                if now >= due_runnable.schedule.get_execution_time():
                    runnables.append(due_runnable)
                    self.scheduled_runnables.pop(0)
                else:
                    break
            return runnables

    async def process_runnables(self, runnables):
        """Process each runnable in turn."""
        Logger.log_info(f"Processing runnable {runnables}")
        for r in runnables:
            await self.process_runnable(r)

    async def process_runnable(self, scheduled_runnable):
        """Process a single runnable."""
        Logger.log_info("Process runnable 1")
        # It's possible that we've been stopped in the interim:
        async with self.lock:
            if self.stopped:
                return
        Logger.log_info("Process runnable 2")

        if scheduled_runnable.runnable.is_valid():
            Logger.log_info("Process runnable 3")
            # Switch to the "main thread" or some main queue for the actual run
            await scheduled_runnable.runnable.run()
            Logger.log_info("Process runnable 4")

            if scheduled_runnable.runnable.stop_dispatcher_after_this_runnable:
                await self.stop()
                return

            # If it must be repeated, schedule it again.
            if scheduled_runnable.schedule.does_runnable_have_to_be_repeated():
                # Recalculate the planned execution time
                scheduled_runnable.schedule.update_execution_time()
                async with self.lock:
                    self.scheduled_runnables.append(scheduled_runnable)
                    self.scheduled_runnables.sort(
                        key=lambda x: x.schedule.get_execution_time()
                    )
           